tutorials/basic-topics/sorting/merge-sort.md

```diff
@@ -11,4 +11,519 @@ keywords:
+    merge_Sort(rightArray)
+
+    i = j = k = 0
+
+    while i < len(leftArray) and j < len(rightArray): 
+        if leftArray[i] < rightArray[j]: 
+            nums[k] = leftArray[i] 
+            i+=1 //incrementing the smaller element side pointer
+        else: 
+            nums[k] = rightArray[j] 
+            j+=1
+        k+=1
+
+    while i < len(leftArray): 
+        nums[k] = leftArray[i] 
+        i+=1
+        k+=1
+
+    while j < len(rightArray): 
+        nums[k] = rightArray[j] 
+        j+=1
+        k+=1
+
+```
+
+</TabItem>
+<TabItem value="java" label="Java">
+
+```java
+// call mergeSort(nums , 0 ,nums.length -1) in public
+
+private void mergeSort(int[] nums, int l, int r) {   // nums -> ArrayList , l -> starting index , r-> end of the ArrayList
+    if (l >= r) return;
+    int mid = l + (r - l) / 2;
+    mergeSort(nums, l, mid);
+    mergeSort(nums, mid + 1, r);
+    merge(nums, l, r);
+}
+private void merge(int[] nums, int l, int r) {
+    int mid = l + (r - l) / 2;
+    int[] tmp = new int[r - l + 1];
+    int i = l, j = mid + 1, k = 0;
+    while (i <= mid || j <= r) {
+        if (i > mid || j <= r && nums[i] > nums[j]) {
+            tmp[k++] = nums[j++];
+        } else {
+            tmp[k++] = nums[i++];
+        }
+    }
+```
+</TabItem>
+</Tabs>
+
+
+## Bottom Up Approach /Iterative technique
+
+### Algorithm 
+
+```
+1. So it  starts with an  **element** in the array 
+
+    It iterative approach it is already considering the fact that one item array is always sorted
+ 
+2. Compares two nearby element to merge a sorted sub array 
+
+    Similarly we merge other sorted subarrays like we have done in top down recursive approach **(two-pointer approach)** 
+
+4. Contines until we a sorted array 
+```
+<Tabs>
+<TabItem value="java" label="Java">
+
+```java
+class Solution {
+    public List<Integer> sortArray(int[] nums) {
+        List<Integer> res = new ArrayList<>();
+        if (nums == null || nums.length == 0) return res;
+        mergeSort2(nums);
+        for (int i : nums) res.add(i);
+        return res;
+    }
+    private void mergeSort2(int[] nums) {
+        for (int size = 1; size < nums.length; size *= 2) {
+            for (int i = 0; i < nums.length - size; i += 2 * size) {
+                int mid = i + size - 1;
+                int end = Math.min(i + 2 * size - 1, nums.length - 1);
+                merge2(nums, i, mid, end);
+            }
+        }
+    }
+    private void merge2(int[] nums, int l, int mid, int r) {
+        int[] tmp = new int[r - l + 1];
+        int i = l, j = mid + 1, k = 0;
+        while (i <= mid || j <= r) {
+            if (i > mid || j <= r && nums[i] > nums[j]) {
+                tmp[k++] = nums[j++];
+            } else {
+                tmp[k++] = nums[i++];
+            }
+        }
+        System.arraycopy(tmp, 0, nums, l, r - l + 1);
+    }
+}
+```
+</TabItem>
+</Tabs>
+
+**Time_Complexity** : Best & Worst & Average is   $O(n \log_{2} n)$
```

Comment:
Why _ ? 
Use Time complexity: and Space complexity: 
Correct : as well.

Reply:
Why undercore ??? remove it

![image](https://user-images.githubusercontent.com/10242660/193453079-42a68412-5fce-4920-8c6a-97d379b59779.png)

Reply:
Still, you havn't removed _ in Time and space complexity. 

Time complexity
Space complexity